### PR TITLE
HDDS-2869. Handle pipeline bootstrap from SCM and create pipeline use case in Recon.

### PR DESCRIPTION
**Pull request by avijayanhwx**
## What changes were proposed in this pull request?

- Implement SCM pipeline metadata bootstrap in Recon.
- Handle CREATE pipeline use case in Recon (Recon sees a new pipeline ID in Datanode heartbeat)
- Stub out pipeline provider in Recon such that it can NEVER create a Ratis pipeline.
- Added SCM RPC endpoint to get a single Pipeline info.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2869

## How was this patch tested?
Manually tested on a 3 node docker cluster.
Added unit tests for new Recon methods.

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineFactory.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
 /**
  * Creates pipeline based on replication type.
  */
-public final class PipelineFactory {
+public class PipelineFactory {
```
Comment (swagle) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineFactory.java]: The access can be made package-private?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java [hunk 2]
```diff
@@ -120,4 +116,24 @@ private void processPipelineReport(PipelineReport report, DatanodeDetails dn,
       }
     }
   }
+
+
+  protected void reportDatanode(Pipeline pipeline, DatanodeDetails dn)
```
Comment (swagle) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java]: For cleaner semantics, this can be renamed to setReportedDatanode.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java [hunk 3]
```diff
@@ -200,6 +210,16 @@ public Pipeline getPipeline(PipelineID pipelineID)
     }
   }
 
+  @Override
+  public boolean containsPipeline(PipelineID pipelineID) {
```
Comment (swagle) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: Instead of contains, we could have made the get return an Optiona<Pipeline>

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java [hunk 4]
```diff
@@ -523,4 +543,21 @@ protected File getPipelineDBPath(Configuration configuration) {
     File metaDir = ServerUtils.getScmDbDir(configuration);
     return new File(metaDir, SCM_PIPELINE_DB);
   }
+
+  protected ReadWriteLock getLock() {
```
Comment (swagle) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: Instead of a get on a lock object, it would make sense to make the reference accessible.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java [hunk 5]
```diff
@@ -408,6 +408,15 @@ public Pipeline createReplicationPipeline(HddsProtos.ReplicationType type,
     return scm.getPipelineManager().getPipelines();
   }
 
+  @Override
+  public Pipeline getPipeline(HddsProtos.PipelineID pipelineID)
+      throws IOException {
+    AUDIT.logReadSuccess(
```
Comment (swagle) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java]: This might not be appropriate for Audit since it is not client access, we should log it in debug mode IMO.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineFactory.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        new ReconMockPipelineProvider();
+    setProviders(new DefaultedMap(reconMockPipelineProvider));
+  }
+
+  static class ReconMockPipelineProvider implements PipelineProvider {
+
+    @Override
+    public Pipeline create(HddsProtos.ReplicationFactor factor){
+      LOG.warn("Trying to create pipeline in Recon, which is prohibited!");
+      return null;
```
Comment (swagle) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineFactory.java]: Should we consider throwing a RuntimeException instead?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineFactory.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  private static final Logger LOG =
+      LoggerFactory.getLogger(ReconPipelineFactory.class);
+
+  ReconPipelineFactory() {
+    ReconMockPipelineProvider reconMockPipelineProvider =
+        new ReconMockPipelineProvider();
+    setProviders(new DefaultedMap(reconMockPipelineProvider));
+  }
+
+  static class ReconMockPipelineProvider implements PipelineProvider {
```
Comment (swagle) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineFactory.java]: I would avoid calling this a Mock since unit tests mocks have a different meaning as they return stubs data structures.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  private OzoneConfiguration configuration;
+  private StorageContainerLocationProtocol scmClient;
+  private boolean isInitialized = false;
+
+  @Inject
+  public StorageContainerServiceProviderImpl(OzoneConfiguration configuration,
+      StorageContainerLocationProtocol scmClient) {
+    this.configuration = configuration;
+    this.scmClient = scmClient;
+    if (this.scmClient != null) {
```
Comment (swagle) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java]: How about calling initialize from the constructor and make that idempotent using the flag.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+ */
+public class StorageContainerServiceProviderImpl
+    implements StorageContainerServiceProvider {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(StorageContainerServiceProviderImpl.class);
+
+  private OzoneConfiguration configuration;
+  private StorageContainerLocationProtocol scmClient;
+  private boolean isInitialized = false;
```
Comment (swagle) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java]: This should be declared volatile.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+      this.scmClient = newContainerRpcClient(configuration);
+    } catch (IOException ioEx) {
+      LOG.error("Exception encountered while creating SCM client.", ioEx);
+    }
+    isInitialized = true;
+  }
+
+  @Override
+  public List<Pipeline> getPipelines() throws IOException {
+    if (!isInitialized) {
```
Comment (swagle) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java]: Same here, and next check becomes redundant.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineFactory.java (hunk 1 above)
Comment (avijayanhwx) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineFactory.java]: This is extended in org.apache.hadoop.ozone.recon.scm.ReconPipelineFactory.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineFactory.java (hunk 6 above)
Comment (avijayanhwx) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineFactory.java]: We don't need to do this since this class protects us against any unintended actions from Recon.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java (hunk 4 above)
Comment (avijayanhwx) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: That will lead to findbug violation. Do you see any issues using a getter?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java (hunk 2 above)
Comment (avijayanhwx) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineReportHandler.java]: Done

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java (hunk 9 above)
Comment (avijayanhwx) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java]: Done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java (hunk 3 above)
Comment (avijayanhwx) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: Since the API is "contains", I wanted to keep it true/false. Can I address this in the next patch?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java (hunk 5 above)
Comment (avijayanhwx) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java]: Done

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java (hunk 3 above)
Comment (swagle) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: It's a minor nit, I am ok with it asis. More of a preference thing, returning Optional already sets the contract and reduces code dup.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java (hunk 4 above)
Comment (swagle) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java]: I see, ok let's ignore this comment.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/audit/SCMAction.java [hunk 11]
```diff
@@ -32,6 +32,7 @@
   GET_CONTAINER_WITH_PIPELINE,
   LIST_CONTAINER,
   LIST_PIPELINE,
+  GET_PIPELINE,
```
Comment (arp7) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/audit/SCMAction.java]: Usually we add new fields at the end of enums.

File: hadoop-ozone/dist/src/main/compose/ozone/docker-compose.yaml [hunk 12]
```diff
@@ -67,6 +67,5 @@ services:
     ports:
       - 9888:9888
     environment:
-      WAITFOR: om:9874
```
Comment (arp7) [inline hadoop-ozone/dist/src/main/compose/ozone/docker-compose.yaml]: Why is this removed?

File: hadoop-ozone/dist/src/main/compose/ozonesecure/docker-compose.yaml [hunk 13]
```diff
@@ -79,7 +79,6 @@ services:
     env_file:
       - ./docker-config
     environment:
-      WAITFOR: om:9874
```
Comment (arp7) [inline hadoop-ozone/dist/src/main/compose/ozonesecure/docker-compose.yaml]: Same.

File: hadoop-ozone/dist/src/main/compose/ozone/docker-compose.yaml (hunk 12 above)
Comment (avijayanhwx) [reply hadoop-ozone/dist/src/main/compose/ozone/docker-compose.yaml]: I just stumbled upon this while testing. We no longer need this. Recon can start up anytime and can learn OM and SCM state.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/audit/SCMAction.java (hunk 11 above)
Comment (avijayanhwx) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/audit/SCMAction.java]: Sure, I can move it to the end.

Comment (arp7) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/audit/SCMAction.java]: I think it is okay for now as we don't have to maintain compatibility.

**Comment by avijayanhwx** [comment]
> Thanks @avijayanhwx for implementing this. I tried it locally on a docker-compose cluster and found that Recon does not get new pipeline information.
> 
> 1. Since now the recon container does not wait with startup, it received 0 pipelines initially:
>    ```
>    recon_1     | 2020-01-29 08:41:05,263 [main] INFO scm.ReconStorageContainerManagerFacade: Obtained 0 pipelines from SCM.
>    recon_1     | 2020-01-29 08:41:05,264 [main] INFO scm.ReconPipelineManager: Recon has 0 pipelines in house.
>    ```
>    
>    
>    That would be OK, but when the pipelines were created a bit later, they did not show up in Recon.  (I also checked with `WAITFOR` restored, 3+1 initial pipelines were received in that case, but I feel it's a bit timing-dependent.)
> 2. Closed the initial 3-node pipeline via `scmcli`, which eventually triggered creation of new one in SCM, but Recon never noticed.
> 
> Can you please check?

@adoroszlai I have not wired up the DN-> Recon path by default since that may cause some unintended test failures before finishing the whole "Recon as a Passive SCM" feature. In my local, I test with the following configs added. 

> OZONE-SITE.XML_ozone.recon.datanode.address=recon:9891
> OZONE-SITE.XML_ozone.recon.address=recon:9891
> OZONE-SITE.XML_ozone.recon.datanode.bind.host=recon 

That will lead to pipelines being picked up by Recon.

**Comment by adoroszlai** [comment]
> I have not wired up the DN-> Recon path by default

Thanks, that explains it.

**Comment by adoroszlai** [comment]
Thanks @avijayanhwx for the contribution, and @arp7 and @swagle for the review.

**Comment by swagle** [comment]
Thanks @adoroszlai for actually trying it out :-) 👍

**Comment by avijayanhwx** [comment]
> Thanks @adoroszlai for actually trying it out :-) 👍

+1. Thanks @adoroszlai
